int32 key
int32 gen_hash
int32 generation
